pkg/providers/instancetype/offering/provider.go

```diff
@@ -0,0 +1,274 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package offering
+
+import (
+	"context"
+	"fmt"
+	"runtime"
+	"sync"
+	"weak"
+
+	ec2types "github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/samber/lo"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/uuid"
+	karpv1 "sigs.k8s.io/karpenter/pkg/apis/v1"
+	"sigs.k8s.io/karpenter/pkg/cloudprovider"
+	"sigs.k8s.io/karpenter/pkg/scheduling"
+
+	v1 "github.com/aws/karpenter-provider-aws/pkg/apis/v1"
+	awscache "github.com/aws/karpenter-provider-aws/pkg/cache"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/pricing"
+)
+
+type Provider interface {
+	InjectOfferings(context.Context, []*cloudprovider.InstanceType, *v1.EC2NodeClass, []string, *types.UID) []*cloudprovider.InstanceType
+}
+
+type DefaultProvider struct {
+	sync.RWMutex
+	snapshots map[types.UID]weak.Pointer[snapshot]
+
+	unavailableOfferings *awscache.UnavailableOfferings
+	pricingProvider      pricing.Provider
+}
+
+func NewDefaultProvider(unavailableOfferingsCache *awscache.UnavailableOfferings, pricingProvider pricing.Provider) *DefaultProvider {
+	return &DefaultProvider{
+		snapshots:            map[types.UID]weak.Pointer[snapshot]{},
+		unavailableOfferings: unavailableOfferingsCache,
+		pricingProvider:      pricingProvider,
+	}
+}
+
+func (p *DefaultProvider) InjectOfferings(
+	instanceTypes []*cloudprovider.InstanceType,
+	nodeClass *v1.EC2NodeClass,
+	allZones sets.Set[string],
+	snapshotUUID *types.UID,
+) []*cloudprovider.InstanceType {
+	subnetZones := lo.SliceToMap(nodeClass.Status.Subnets, func(s v1.Subnet) (string, string) {
+		return s.Zone, s.ZoneID
+	})
+	its := []*cloudprovider.InstanceType{}
+	for _, it := range instanceTypes {
+		offerings := p.createOfferings(
+			it,
+			nodeClass,
+			allZones,
+			subnetZones,
+			snapshotUUID,
+		)
+		for _, of := range offerings {
+			InstanceTypeOfferingAvailable.Set(float64(lo.Ternary(of.Available, 1, 0)), map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+			InstanceTypeOfferingPriceEstimate.Set(of.Price, map[string]string{
+				instanceTypeLabel: it.Name,
+				capacityTypeLabel: of.Requirements.Get(karpv1.CapacityTypeLabelKey).Any(),
+				zoneLabel:         of.Requirements.Get(corev1.LabelTopologyZone).Any(),
+			})
+		}
+
+		its = append(its, &cloudprovider.InstanceType{
```

Comment:
So this just creates a new struct container for the same data -- so it's not affecting the original, just adding a new container that contains the offerings, do I have that right? Do we have a feeling what the memory cost for this is? Given that this is just adding another pointer that should be ref-ing out to other pointers, it's probably not very expensive, but just validating

Reply:
It's going to be ~50 kilobytes extra per call to `GetInstanceTypes`, shouldn't be too significant in the scheme of things.